### PR TITLE
Kde test suite

### PR DESCRIPTION
**Pull request by AlexanderFengler**

### CODE REVIEW

File: notebooks/kde_class.ipynb [hunk 1]
```diff
@@ -0,0 +1,142 @@
+{
```
Comment (cpaniaguam) [inline notebooks/kde_class.ipynb:1]: <p>Does this bit run okay?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/lnccbrown/ssm-simulators/pull/94/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

File: ssms/support_utils/kde_class.py [hunk 2]
```diff
@@ -513,297 +455,47 @@ def attach_data_from_simulator(
...
         when generating gaussian-based kdes in the LogKDE class.
     """
     # Compute number of samples
-    n = len(sample)
+    sample_ = np.array(sample)
+    n = len(sample_)
 
     # Deal with very small stds and n = 1 case
     if n > 1:
         # Compute std of sample
```
Comment (cpaniaguam) [inline ssms/support_utils/kde_class.py:490]:
```suggestion
```

File: ssms/support_utils/kde_class.py [hunk 3]
```diff
@@ -513,297 +455,47 @@ def attach_data_from_simulator(
...
-        std = np.std(sample)
+        std = np.std(sample_)
         if std < std_cutoff:
             if std_proc == "restrict":
                 std = std_cutoff
             if std_proc == "kill":
-                std = 0
+                std = 0.0
```
Comment (cpaniaguam) [inline ssms/support_utils/kde_class.py:496]:
```suggestion
        std = max(std, std_cutoff) if std_proc == "restrict" else 0.0
```

File: ssms/support_utils/kde_class.py [hunk 4]
```diff
@@ -513,297 +455,47 @@ def attach_data_from_simulator(
...
     else:
         # AF-Comment: This is a bit of a weakness (can be arbitrarily incorrect)
         std = std_n_1
 
     return np.power((4 / 3), 1 / 5) * std * np.power(n, (-1 / 5))
```
Comment (cpaniaguam) [inline ssms/support_utils/kde_class.py:501]:
```suggestion
    return np.power((4 / (3*n)), 1 / 5) * std
```

File: tests/test_kde_class.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    assert kde.auto_bandwidth is True
+    assert kde.bandwidth_type == "silverman"
+
+
+def test_bandwidth_silverman_with_std():
+    """Test bandwidth_silverman function."""
+    sample = np.array([1.0, 2.0, 3.0])
+    bw = bandwidth_silverman(sample)
+    assert isinstance(bw, float)
+    assert bw > 0
```
Comment (cpaniaguam) [inline tests/test_kde_class.py:28]: Are there edge cases to consider?

File: tests/test_kde_class.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    result = kde.kde_eval(eval_data, lb=-66.774)
+    assert isinstance(result, np.ndarray)
+    assert len(result) == 2
+    assert np.all(result >= -66.774)
+
+
+def test_kde_sample(sample_ddm_data):
+    """Test kde_sample method."""
+    kde = LogKDE(simulator_data=sample_ddm_data)
+    samples = kde.kde_sample(n_samples=100)
```
Comment (cpaniaguam) [inline tests/test_kde_class.py:77]: Can the actual values in the sample be checked if the random seed is set?

**Review by cpaniaguam — CHANGES_REQUESTED** [review]
Thanks for this, Alex! Left a number of suggestions for your consideration.

File: notebooks/kde_class.ipynb (hunk 1 above)
Comment (AlexanderFengler) [reply notebooks/kde_class.ipynb:1]: notebook is not really ready tbh

File: tests/test_kde_class.py (hunk 6 above)
Comment (AlexanderFengler) [reply tests/test_kde_class.py:77]: turned this thing into a separate issue, will include this in the context of finishing the notebook.

File: tests/test_kde_class.py (hunk 5 above)
Comment (AlexanderFengler) [reply tests/test_kde_class.py:28]: should always be positive on simple call. 
added extra `ValueError` if empty sample gets passed.

**Comment by AlexanderFengler** [comment]
@cpaniaguam should be ready for re-review.

File: notebooks/kde_class.ipynb (hunk 1 above)
Comment (AlexanderFengler) [reply notebooks/kde_class.ipynb:1]: running now

Comment (AlexanderFengler) [reply notebooks/kde_class.ipynb:1]: albeit pretty sparse notebook, I will work on narrative and put this properly in the docs later.

File: ssms/support_utils/kde_class.py [hunk 7]
```diff
@@ -184,182 +194,107 @@ def kde_eval(self, data={}, log_eval=True, lb=-66.774, eps=10e-5, filter_rts=-99
...
-                )  # -66.77497 # the number corresponds to log(1e-29)
-                # --> should rather be log(1 / n) + log(1 / 20)
+                ) + np.log(1 / self.simulator_info["max_t"])
             else:
+                # Evaluate likelihood for "choices" for which we have base data
                 log_kde_eval_out_tmp = log_rts[choice_idx_tmp]
-                log_kde_eval_out_tmp[displaced_rts[choice_idx_tmp] <= 0] = lb
-                log_kde_eval_out_tmp[displaced_rts[choice_idx_tmp] > 0] = (
+                print("BEFORE")
+                print(log_kde_eval_out_tmp)
```
[the remark this answers is not in the document]
Comment (AlexanderFengler) [reply ssms/support_utils/kde_class.py]: dropped this but yes otherwise should be using logging everywhere

File: ssms/support_utils/kde_class.py [hunk 8]
```diff
@@ -184,182 +194,107 @@ def kde_eval(self, data={}, log_eval=True, lb=-66.774, eps=10e-5, filter_rts=-99
...
-                log_kde_eval_out_tmp[displaced_rts[choice_idx_tmp] > 0] = np.log(
-                    self.data["choice_proportions"][self.data["choices"].index(c)]
-                ) + self.base_kdes[self.data["choices"].index(c)].score_samples(
-                    np.expand_dims(
-                        log_rts[choice_idx_tmp][displaced_rts[choice_idx_tmp] > 0], 1
-                    )
-                )
                 log_kde_eval[choice_idx_tmp] = log_kde_eval_out_tmp
 
         if log_eval is True:
```
[the remark this answers is not in the document]
Comment (AlexanderFengler) [reply ssms/support_utils/kde_class.py:298]: done

File: ssms/support_utils/kde_class.py [hunk 9]
```diff
@@ -368,34 +303,40 @@ def _kde_eval_log_rt(
...
-        # kde_samples = []
-
         rts = np.zeros((n_samples, 1))
         choices = np.zeros((n_samples, 1))
 
-        n_by_choice = []
+        n_by_choice: list[int] = []
         for i in range(0, len(self.data["choices"]), 1):
             if use_empirical_choice_p is True:
                 n_by_choice.append(
```
[the remark this answers is not in the document]
Comment (AlexanderFengler) [reply ssms/support_utils/kde_class.py]: improved

File: ssms/support_utils/kde_class.py [hunk 10]
```diff
@@ -513,297 +455,47 @@ def attach_data_from_simulator(
...
-    sample=[  # noqa: B006
-        0,
-        0,
-        0,
-    ],  # TODO: #81 B008 Do not perform function call `np.arange` in argument defaults; instead, perform the call within the function, or read the default from a module-level singleton variable  # noqa: B006, B008, FIX002
-    std_cutoff=1e-3,
-    std_proc="restrict",  # options 'kill', 'restrict'
-    std_n_1=10,  # HERE WE CAN ALLOW FOR SOMETHING MORE INTELLIGENT
-):
+    sample: np.ndarray | list[float] | tuple[float] = (0, 0, 0),
```
[the remark this answers is not in the document]
Comment (AlexanderFengler) [reply ssms/support_utils/kde_class.py]: done

**Comment by AlexanderFengler** [comment]
@cpaniaguam should be good for real now :)

File: ssms/support_utils/kde_class.py [hunk 11]
```diff
@@ -183,226 +190,166 @@ def kde_eval(self, data={}, log_eval=True, lb=-66.774, eps=10e-5, filter_rts=-99
...
-            if use_empirical_choice_p is True:
-                n_by_choice.append(
-                    round(n_samples * self.data["choice_proportions"][i])
-                )
-            else:
-                n_by_choice.append(round(n_samples * alternate_choice_p[i]))
+        if (
+            not len(alternate_choice_p) == len(self.data["choices"])
+            and not use_empirical_choice_p
+        ):
```
Comment (cpaniaguam) [inline ssms/support_utils/kde_class.py:340]:
```suggestion
        if not any(
            [
                len(alternate_choice_p) == len(self.data["choices"]),
                use_empirical_choice_p,
            ]
        ):
```

File: ssms/support_utils/kde_class.py [hunk 12]
```diff
@@ -140,23 +142,28 @@ def generate_base_kdes(self, auto_bandwidth=True, bandwidth_type="silverman"):
...
+        self.base_kdes = [
+            (
+                "no_base_data"
+                if self.bandwidths[i] == "no_base_data"
+                else KernelDensity(kernel="gaussian", bandwidth=self.bandwidths[i]).fit(
+                    np.log(self.data["rts"][i])
                 )
+            )
+            for i in range(len(self.data["choices"]))
+        ]
```
[the remark this answers is not in the document]
Comment (AlexanderFengler) [reply ssms/support_utils/kde_class.py:157]: keeping the loop over `choices` for ease of comprehension but adjusted the rest
